File: starter-kit/typescript/agent.ts
/**
 * Poker Tournament — TypeScript Starter Agent
 *
 * Connects to the game server via WebSocket and plays random legal actions.
 * Replace the `decide()` function with your own logic.
 *
 * Setup: cp .env.example .env  then edit .env
 * Run:   npm start
 */

import 'dotenv/config'
import WebSocket from 'ws'

const SERVER_URL  = process.env.POKER_SERVER  ?? 'ws://localhost:3000'
const AGENT_ID    = process.env.AGENT_ID      ?? 'ts-agent-1'
const AGENT_NAME  = process.env.AGENT_NAME    ?? 'TypeScriptBot'

// ── Types (mirrors protocol.ts on the server) ───────────────────────────────

interface Card { rank: number; suit: 'c' | 'd' | 'h' | 's' }

interface ActionRequiredMsg {
  type: 'action_required'
  gameId: string
  handNumber: number
  holeCards: [Card, Card]
  communityCards: Card[]
  pot: number
  myStack: number
  myBet: number
  currentBet: number
  validActions: string[]
  minRaise: number
  maxRaise: number
}

interface HandResultMsg {
  type: 'hand_result'
  handNumber: number
  winners: { playerId: string; amount: number }[]
}

interface TournamentUpdateMsg {
  type: 'tournament_update'
  standings: { playerId: string; stack: number; eliminated: boolean }[]
  blindLevel: number
  smallBlind: number
  bigBlind: number
}

type ServerMsg = ActionRequiredMsg | HandResultMsg | TournamentUpdateMsg | { type: 'error'; message: string }

// ── Decision logic (replace this!) ──────────────────────────────────────────

function decide(state: ActionRequiredMsg): { action: string; amount?: number } {
  const { validActions, minRaise, maxRaise } = state

  // Never fold if we can check
  if (validActions.includes('CHECK')) return { action: 'CHECK' }

  // Raise 20% of the time
  if (validActions.includes('RAISE') && Math.random() < 0.2) {
    const amount = Math.floor(minRaise + Math.random() * (Math.min(minRaise * 3, maxRaise) - minRaise))
    return { action: 'RAISE', amount }
  }

  if (validActions.includes('CALL')) return { action: 'CALL' }

  return { action: 'FOLD' }
}

// ── Connection loop ──────────────────────────────────────────────────────────

function connect() {
  console.log(`Connecting to ${SERVER_URL} as ${AGENT_NAME} (${AGENT_ID})`)
  const ws = new WebSocket(SERVER_URL)

  ws.on('open', () => {
    ws.send(JSON.stringify({ type: 'register', agentId: AGENT_ID, agentName: AGENT_NAME }))
    console.log('Registered. Waiting for hands...')
  })

  ws.on('message', (raw) => {
    const msg = JSON.parse(raw.toString()) as ServerMsg

    if (msg.type === 'action_required') {
      const action = decide(msg)
      ws.send(JSON.stringify({ type: 'action', gameId: msg.gameId, ...action }))
    }

    if (msg.type === 'hand_result') {
      const win = msg.winners.find(w => w.playerId === AGENT_ID)
      if (win) console.log(`Won hand #${msg.handNumber}! +${win.amount}`)
    }

    if (msg.type === 'tournament_update') {
      const me = msg.standings.find(p => p.playerId === AGENT_ID)
      if (me) console.log(`Stack: ${me.stack.toLocaleString()}  |  Blinds ${msg.smallBlind}/${msg.bigBlind}`)
    }

    if (msg.type === 'error') {
      console.error(`Server error: ${msg.message}`)
    }
  })

  ws.on('close', () => {
    console.log('Disconnected. Reconnecting in 2s...')
    setTimeout(connect, 2000)
  })

  ws.on('error', (err) => {
    console.error('WebSocket error:', err.message)
  })
}

connect()

File: starter-kit/python/agent.py
"""
Poker Tournament — Python Starter Agent

Connects to the game server via WebSocket and plays random legal actions.
Replace the `decide()` function with your own logic.

Setup: cp .env.example .env  then edit .env
Run:   python agent.py
"""

import asyncio
import json
import os
import random
import websockets
from dotenv import load_dotenv

load_dotenv()

SERVER_URL = os.environ.get("POKER_SERVER", "ws://localhost:3000")
AGENT_ID   = os.environ.get("AGENT_ID", "python-agent-1")
AGENT_NAME = os.environ.get("AGENT_NAME", "PythonBot")


def decide(state: dict) -> dict:
    """
    Return an action given the current game state.

    state keys:
      holeCards       – list of 2 cards [{"rank": int, "suit": str}]
      communityCards  – list of 0–5 cards
      pot             – int
      myStack         – int
      currentBet      – int
      myBet           – int
      validActions    – list of "FOLD" | "CHECK" | "CALL" | "RAISE"
      minRaise        – int
      maxRaise        – int

    Return one of:
      {"action": "FOLD"}
      {"action": "CHECK"}
      {"action": "CALL"}
      {"action": "RAISE", "amount": <int>}
    """
    valid = state["validActions"]

    # Example: never fold if check is available; otherwise call; raise 20% of the time
    if "CHECK" in valid:
        return {"action": "CHECK"}

    if "RAISE" in valid and random.random() < 0.2:
        amount = random.randint(state["minRaise"], min(state["minRaise"] * 3, state["maxRaise"]))
        return {"action": "RAISE", "amount": amount}

    if "CALL" in valid:
        return {"action": "CALL"}

    return {"action": "FOLD"}


async def run():
    print(f"Connecting to {SERVER_URL} as {AGENT_NAME} ({AGENT_ID})")
    async with websockets.connect(SERVER_URL) as ws:
        # Register with the server
        await ws.send(json.dumps({
            "type": "register",
            "agentId": AGENT_ID,
            "agentName": AGENT_NAME,
        }))
        print("Registered. Waiting for hands...")

        async for raw in ws:
            msg = json.loads(raw)

            if msg["type"] == "action_required":
                action = decide(msg)
                await ws.send(json.dumps({
                    "type": "action",
                    "gameId": msg["gameId"],
                    "action": action["action"],
                    "amount": action.get("amount"),
                }))

            elif msg["type"] == "hand_result":
                for w in msg.get("winners", []):
                    if w["playerId"] == AGENT_ID:
                        print(f"Won hand #{msg['handNumber']}! +{w['amount']}")

            elif msg["type"] == "tournament_update":
                me = next((p for p in msg["standings"] if p["playerId"] == AGENT_ID), None)
                if me:
                    print(f"Stack: {me['stack']:,}  |  Blind level {msg['blindLevel']}: {msg['smallBlind']}/{msg['bigBlind']}")

            elif msg["type"] == "error":
                print(f"Server error: {msg['message']}")


if __name__ == "__main__":
    asyncio.run(run())
